WikiWatershed/model-my-watershed
# ITSI Registration front-end

Adds a login and registration front-end for users logging in with ITSI.

**Before You Begin**

Please check if the `MMW_ITSI_SECRET_KEY` variable is set on your `app` VM:

    vagrant ssh app -c 'cat /etc/mmw.d/env/MMW_ITSI_SECRET_KEY'

If it isn't, you must reprovision the `app` VM with the `MMW_ITSI_SECRET_KEY` variable set, like so:

    MMW_ITSI_SECRET_KEY=*** vagrant provision app

The secret key is available in the wiki.

**Testing Instructions**

 1. Open [http://localhost:8000/](http://localhost:8000/). You should see a new login page that matches the latest mockups.
 2. Click on "ITSI". You should be taken to the ITSI portal.
 3. Log in with either the `tstudent2` or `mmwterence` account to ITSI (credentials are available in the wiki). You should be redirected to the registration front-end, and the `username`, `first_name`, and `last_name` fields should be pre-populated.
 4. Check "I agree to something" and click "Sign Up". If there aren't any validation errors, the registration modal should dismiss and you should be logged in, apparent from the username in the top right hand corner.
 5. Refresh the page to check that you are still logged in as this newly created user.
 6. Log out of this account. Click "Login", and again click "ITSI". You will be redirected to the ITSI portal again, but since you recently logged in there you should be redirected right back to our app which should automatically log you in.

**Alternate Paths**

 * Try going to [`/register`](http://localhost:8000/register) directly and registering.
 * Try dismissing the modal during registration.
 * Try signing up as an already existing username.

**Notes**

 * The mockup has a line from edge to edge behind the "Or continue as" text which I don't know how to add:  
![screen shot 2015-06-02 at 4 28 32 pm](https://cloud.githubusercontent.com/assets/1430060/7946427/93c1ce52-0944-11e5-817c-8de6eb240a54.png)
 * We pass back the username, first and last names received from ITSI to the registration modal via the query string. Is this a security / aesthetic issue? Should we use cookies or some other "under the hood" mechanism which is uglier in code but cleaner in the UI?
 * Unlike regular sign-up / login, the ITSI login requires full page refreshes and is not an XHR-only operation. This should be kept in mind when creating mechanisms to save temporary work in case the user decides to create an account using ITSI mid-way through an unsaved project.
 * We now have a handful of modal views that are very similar in basic operation. Perhaps we should refactor the commonalities into a utility class and derive from that instead.
 * I needed to add some styling for the login modal updates. I decided to add it under `sass/base/_login.scss` and not `sass/pages/_login.scss` because the login is a modal, not a page, and would be available throughout the app, not just in a particular view. If this was not the right place to add it, please let me know and I will correct it.

Connects #180
